src/NUnitFramework/framework/Assert.cs

```diff
@@ -324,16 +336,19 @@ private static void ReportFailure(ConstraintResult result, string message, param
         {
             MessageWriter writer = new TextMessageWriter(message, args);
             result.WriteMessageTo(writer);
-            string formattedMessage = writer.ToString();
-            string stackTrace = GetStackTrace();
 
+            ReportFailure(writer.ToString());
+        }
+
+        private static void ReportFailure(string message)
+        {
             // Failure is recorded in <assertion> element in all cases
             TestExecutionContext.CurrentContext.CurrentResult.RecordAssertion(
-                AssertionStatus.Failed, formattedMessage, stackTrace);
+                AssertionStatus.Failed, message, GetStackTrace());
```

Comment:
It turns out this may cause a problem in some situations... When Assert.Throws is used to run code that may contain an assertion, the nested assertion ends up adding failures to the result. When the Throws succeeds, those lower level failures are still present in the result. 

This problem is already present in master, due to an earlier merge, so I would prefer to finish up this PR and write a separate issue for the problem, which is fairly obscure and easier to work on in isolation.

```
Assert.Throws<